R/prepare_stop_level_data.R

```diff
@@ -26,146 +29,347 @@ granulate_to_stop <- function(mbbs) {
+  mbbs %>%
+    # Keep rows that have non-blank species_comments
+    filter(species_comments != "") %>%
+    # Keep rows that aren't already separated by stop
+    filter(is.na(stop_num)) %>%
+    # Keep rows that contain at least one number
+    filter(str_detect(species_comments, "[0-9]")) %>%
+    # Remove pre-dawn owling checklist
+    filter(sub_id != "S46296942") %>%
+    
+    mutate(species_comments = fix_species_comments(species_comments)) %>%
+    # Remove rows that were changed to have blank species_comments
+    filter(species_comments != "") %>%
+    dplyr::bind_cols(
+      dplyr::tibble(
+        !!! rlang::set_names(rep(NA_character_, 21), 
+                             c(paste0("s", 1:20), "sc_note"))
+      )
+    ) %>%
+    #set s1:s20 to integer columns
+    mutate(across(.cols = s1:s20, .fns = as.integer)) %>%
+    # relocate to be the first columns is necessary b/c
+    # later data is added to a given [row,column] based on columns 1:20. 
+    # Also helpful for temp and error identification
+    relocate(s1:s20, species_comments, count, sc_note)
+}
+
+
+#'
+#'
+#' @param stopsmbbs a row from the mbbs that's gone through the prepare_to_process
+#'      function, ie: stopsmbbs. MUST have $species_comments, $count, and $s1:s20
+direct_to_sp_com_function <- function(stopsmbbs) {
+  
+  assertthat::assert_that(
+    is.character(stopsmbbs$species_comments),
+    is.numeric(stopsmbbs$count),
+    nrow(stopsmbbs) == 1
+  )
+  
+  #initiate stop counts list
+  stop_counts_list <- NA
+  
+  #route species comments to the relevent function to process it.
+  if(str_starts(stopsmbbs$species_comments, ",|[0-9]+,")) {
+    
+    stop_counts_list <- 
+      sp_com_comma_seperated(stopsmbbs$species_comments, stopsmbbs$count)
+    
+  } else if (str_starts(stopsmbbs$species_comments, "[0-9]+=[0-9]+|st(op)?( )?[0-9]+=[0-9]+")) {
+    
+    stop_counts_list <- 
+      sp_com_stop_equals_count(stopsmbbs$species_comments, stopsmbbs$count)
+    
+  } else if (str_starts(stopsmbbs$species_comments, "st(op)?")) {
     
-  #we need to add the rows for imputing the stop information to stopsmbbs
-  stopsmbbs <- stopsmbbs %>%
-    mutate("s1" = NA, 's2' = NA,'s3' = NA,'s4' = NA,"s5" = NA,"s6" = NA,"s7" = NA,"s8" = NA,"s9" = NA,"s10" = NA,"s11" = NA,"s12" = NA,"s13" = NA,"s14" = NA,"s15" = NA,"s16" = NA,"s17" = NA,"s18" = NA,"s19" = NA,"s20" = NA, 
-           "sc_note" = NA) %>%
-    relocate(s1:s20, species_comments, count, sc_note) #relocate to be the first columns is necessary b/c later data is added to a given [row,column] based on columns 1:20. Also helpful for temp and error identification
-
-  #set up for-loop to fill in s1:s20 with information from species_commentts
-  split_list <- NA; stop <- NA; count <- NA
-  #for-loop 
+    stop_counts_list <- 
+      sp_com_only_stop(stopsmbbs$species_comments, stopsmbbs$count)
+    
+  }
+  
+  #add the new list to s1:s20 of stopsmbbs
+  for(i in 1:length(stop_counts_list)) { #which will always be 20
```

Comment:
Maybe I'm misunderstanding, but doesn't this replace the `i`th column of `stopsmbbs` with the content of the `i`th element of `stop_counts_list`? How do you know that (e.g.) `stopsmbbs[ , 1]` corresponds to `s1`?

Reply:
If `stops_counts_list` is a `list` and you want a `data.frame` structure, you could do something like:

```r
x <- list(3 , 4, 6) # just some example data
dplyr::as_tibble(setNames(x, paste0("s", seq_len(length(x)))))
```

Reply:
#make stop_counts_list into a df
  stop_counts_df <- 
    data.frame(scl = stop_counts_list, snum = paste0("s", 1:20)) %>%
    pivot_wider(names_from = snum, values_from = scl)

 #merge with the stopsmbbs_row
  stopsmbbs_row <- 
    stopsmbbs_row %>%
    mutate(stop_counts_df)

Reply:
This would work great, and remove the need to specifically set-up s1:s20 in the prepare_to_process function, but the problem I'm running into is then returning this row and replacing it within the dataset. Totally possible this would not be a problem if I could get purrr::map working but I really don't want to spend more time on trouble-shooting that. In the for(1 in nrow(stopsmbbs)) I have now, I'd have to like create a temp and then rbind new rows to the temp... not prefered. Thinking out loud through these comments, may come to a solution in just a minute

Reply:
Alt that instead of replacing the [i]th column (risky, what if order changes?) does specify s1 gets stop_count_list[1]:

#  s <- paste0("s", 1:20)
 # for(i in 1:length(stop_counts_list)) { #which will always be 20
#    stopsmbbs_row <- stopsmbbs_row %>%
#      mutate(!!as.name(s[i]) := stop_counts_list[i])
#  }

Just requires that part of prepare_to_process continues to be creating the columns s1:s20